### ISSUE TITLE
Move to a single d2-ui package

### ISSUE THREAD

**Issue by jenniferarnesen** [dhis2/notes#20]
It seems like we are leaning towards moving to a single d2-ui npm package with named exports for each of the components rather than components being maintained in separate packages.

If it is decided to go for this, how should it be done?

**Comment by jenniferarnesen** [comment]
Here is a suggestion for a soft migration:

Add a `d2-ui` package in d2-ui, published as @dhis2/d2-ui. Start by putting the new ItemSelector component in there as a named export. Every time a developer is going to make a change to an existing d2-ui component, they move it to packages/d2-ui and DEPRECATE the existing npm package. Then delete the standalone package. Any remaining standalone packages that use the migrated component will need their dependency updated.

**Comment by Mohammer5** [comment]
Side note: I don't know if this is already in place, but we def. need to keep an eye on this:
https://webpack.js.org/guides/tree-shaking/#mark-the-file-as-side-effect-free

**Comment by varl** [comment]
I think it helps to sketch out what-goes-where a bit, this is not set in stone but it does have a way out of **backporting Hell** for the components.

### `dhis2/ui`

- Prop-based primitive components and widgets (global)
- No dependencies
- Follows semver

```
dhis2/ui
├── core
│   └── Button
├── css
├── utils
└── widgets
    └── HeaderBar
```

### `dhis2/d2-ui`

- Contains drivers for different API versions (e.g. HOCs, Hooks, Containers)
- Can utilize external dependencies
- Follows semver
- No backporting - always move forward
- Specialized components go here

```
dhis2/d2-ui
├── OrgUnitTree
├── PeriodSelector
└── HeaderBar
    ├── driver/index.js      # contains default driver for api (if they haven't changed)
    ├── v30/index.js         # contains driver for v30 api
    ├── v31/index.js         # adds new driver for v31 api
    └── index.js
```

The drivers would either contain a [`Container` element](https://github.com/dhis2/ui/blob/5c98df67505c9c78cd7e76ef55b3d5fe2b8ab54a/src/widgets/HeaderBar/HeaderBar.Container.js) or wrap the `dhis2/ui` primitive in some way (e.g. Hook, HOC) which does the API requests and passes props to the primitive component.

If we can keep the library moving forward without having to upgrade _all_ components at the same time in d2-ui, then it is a lot simpler to keep them in the same package.

#### Which modules do we export?
In `d2-ui-components` today we only export the ES module builds, and this once again (_argh_) requires us to use `transformIgnorePatterns` in Jest to allow the builds to work.

In `d2-ui` we transpile ESM to CJS modules (strictly speaking), but Babel adds a property to the exported object [`__esModule`](https://github.com/d2-ci/ui/blob/master/core/Button/index.js#L3-L5) which allows Webpack to treat it as a ESM at bundle-time.

#### Shake, shake, shake -- _shake it off_
We have verified that it is possible to tree-shake using transpiled modules (ESM => CJS) as long as you do not import the re-export file.

This is tree-shakeable: [`import Button from '@dhis2/ui/core/Button'`](https://github.com/d2-ci/ui/blob/master/core/Button/index.js)
This is **not** tree-shakeable: [`import { Button } from '@dhis2/ui'`](https://github.com/d2-ci/ui/blob/master/core/index.js)

#### Up to the apps

Some apps might have a mechanism for centrally controlling all the API calls, so they would use the pure primitives from `dhis2/ui`, while other apps which do not want to bother with that, could import the wrapped component from `@dhis2/d2-ui`.

#### Soft or hard migration

I think both works. In the case of a soft migration, cleaning up the transpile/build scripts would come at the end. And updating all the dependencies as well. Maybe that's a good thing, but it means we'll base the builds of the new `d2-ui` package on the old build settings for a while.

In the case of a soft migration the [`deploy-build`](https://github.com/dhis2/deploy-build) scripts needs to be adjusted, otherwise it will mangle the d2-ui package name.

**Comment by jenniferarnesen** [comment]
> dhis2/d2-ui
> ├── OrgUnitTree
> ├── PeriodSelector
> └── HeaderBar
>     ├── driver/index.js      # contains default driver for api (if they haven't changed)
>     ├── v30/index.js         # contains driver for v30 api
>     ├── v31/index.js         # adds new driver for v31 api
>     └── index.js

I like this solution

**Comment by Mohammer5** [comment]
@varl That's a great overview! And looks like it's kind of complete (although being in draft stage).
We should copy that somewhere as docs, this is definitely quite useful.

I like the structure!

**Comment by ismay** [comment]
> It seems like we are leaning towards moving to a single d2-ui npm package with named exports for each of the components rather than components being maintained in separate packages.

I like that. It seems that the packages were quite tightly coupled, so putting them in a single package makes more sense to me than publishing them separately.

As for the name, it's a nitpick, but `d2-ui` only makes sense to us as a team, or those familiar with the `d2` library. To outsiders it doesn't immediately mean something, and the `d2` part of the name is basically an implementation detail right? To me something like `dhis2/ui` & `@dhis2/stateful-ui` would make more sense if I were new to our ecosystem. Plus if we remove `d2` as a dependency in favor of `d2-api` the name will still make sense.

```
dhis2/d2-ui
├── OrgUnitTree
├── PeriodSelector
└── HeaderBar
    ├── driver/index.js      # contains default driver for api (if they haven't changed)
    ├── v30/index.js         # contains driver for v30 api
    ├── v31/index.js         # adds new driver for v31 api
    └── index.js
```

I do find it difficult to predict how the driver approach would work out. The volatility of our api is not a problem I've dealt with on this scale before, so I'm wondering if there's any other organisations that had to deal with a similar problem, that we could maybe learn from? Other than that I would love it if this would solve our backporting problem. I do like the abstraction between the component and the api version, and the colocation of it.

Also, I was thinking that eventually I can imagine us publishing a `dhis2/hooks` library without any ui. Or maybe we'll export the hooks separately from `d2-ui`. It does seems nice to decouple the hooks completely from the ui.

It's too bad that the tree shaking doesn't allow us to export everything from a main entry point by the way. I'm seeing a lot of deep imports from d2 in the d2-ui library, and it completely ties d2-ui to d2's folder structure, where that should be private to the module. But I guess that we'll just need to accept it for the benefits of tree shaking.

**Comment by ismay** [comment]
> It's too bad that the tree shaking doesn't allow us to export everything from a main entry point by the way. I'm seeing a lot of deep imports from d2 in the d2-ui library, and it completely ties d2-ui to d2's folder structure, where that should be private to the module. But I guess that we'll just need to accept it for the benefits of tree shaking.

Maybe, at the very least, we could have entry points reexporting each component separately, so that we're not exporting from the internal folder structure of the module, whilst still enabling treeshaking.

**Comment by varl** [comment]
>To me something like dhis2/ui & @dhis2/stateful-ui would make more sense if I were new to our ecosystem. 

That's a good point, it doesn't need to be named `d2-ui` at all. This is an opportunity to come up with something else as not all components will have a `d2` dependency and we are trying to walk away from having `d2-` as a prefix on everything.

**Comment by Mohammer5** [comment]
> This is tree-shakeable: [`import Button from '@dhis2/ui/core/Button'`](https://github.com/d2-ci/ui/blob/master/core/Button/index.js)
> This is **not** tree-shakeable: [`import { Button } from '@dhis2/ui'`](https://github.com/d2-ci/ui/blob/master/core/index.js)

@varl 
You said that you've tested this, but in my opinion, this should be possible (lodash get it done somehow). Maybe `"sideEffects": false` mit be of help here? (They [do that](https://github.com/lodash/lodash/blob/e7221b96ec2073e578e948966f0fce8815815c36/package.json#L10) as well)

> In a 100% ESM module world, identifying side effects is straightforward. However, we aren't there just yet, so in the mean time it's necessary to provide hints to webpack's compiler on the "pureness" of your code.
> 
> The way this is accomplished is the "sideEffects" package.json property.
> ([https://webpack.js.org/guides/tree-shaking/#mark-the-file-as-side-effect-free](https://webpack.js.org/guides/tree-shaking/#mark-the-file-as-side-effect-free))

@ismay @varl 
> To me something like dhis2/ui & @dhis2/stateful-ui would make more sense if I were new to our ecosystem.

I like this idea, what about
* ui-stateless
* ui-isolated (or ui-stateful)

I'd interpret `stateful` as react components that are classes, not function. I still wouldn't know that it'll fetch the backend directly, which is why I thought maybe `isolated` makes more sense here? Not so sure thought

**Comment by varl** [comment]
> You said that you've tested this, but in my opinion, this should be possible (lodash get it done somehow). Maybe `"sideEffects": false` mit be of help here? (They [do that](https://github.com/lodash/lodash/blob/e7221b96ec2073e578e948966f0fce8815815c36/package.json#L10) as well)

Lodash produces [different packages for different environments](https://github.com/lodash/lodash/tree/e7221b96ec2073e578e948966f0fce8815815c36#module-formats), and their `lodash-es` package is straight up ESM: https://github.com/lodash/lodash/blob/4.17.11-es

We are using the `sideEffects` prop in `dhis2/ui` as well, but I think the root of it is that we transpile ESM=>CJS, so there are some limitations surrounding that.

This is a bit off topic though.

**Comment by varl** [comment]
@jenniferarnesen sorry about the side-track, and back to the topic...

If you create a `d2-ui/packages/d2-ui` package, the scripts are going to deploy that to d2-ci as `d2-ui-d2-ui`, which is funny but not that helpful. :clown_face: 

One way of doing this technically would be to modify the [`deploy-build.sh`](https://github.com/dhis2/deploy-build/blob/master/deploy-build.sh#L202) script to drop the prefix, and get the `name` from `package.json` instead of building it up from assumptions made about the filesystem. We would have to strip the scope from the name string, but that's fine.

After that point you should be good to create `d2-ui/packages/d2-ui` and have it function in the build chain. I think that should be enough to get started.

**Comment by varl** [comment]
I think it's unfair that the `dhis2/ui` library squats the `widgets` name internally, we could flatten that package completely and rename:

- `ui` becomes `ui-core`
- `d2-ui` becomes `ui-widgets`

**Comment by varl** [comment]
@jenniferarnesen I've updated the build scripts so you could use the name `@dhis2/ui-widgets` from your new package and it will be deploy and published properly.

**Comment by jenniferarnesen** [comment]
Overview of d2 dependency of d2-ui components:

**components not dependent on d2:**

* rich-text
* item-selector (new, in branch)
* formula-editor (d2 is listed in package.json, but I do not find any actual usage)

**components that depend on d2 but don't need to:**

* table (d2 used for translations and getting user locale)
* forms (d2 used for translations only)
* group-editor (d2 used for translations only)
* icon-picker (d2 used for translations only)

**components dependent on d2:**

* expression-manager
* header-bar
* interpretations
* legend
* file-menu
* translation-dialog
* legend
* mentions-wrapper
* org-unit-dialog
* org-unit-select
* org-unit-tree
* period-selector-dialog
* sharing-dialog
* translations-dialog

**Comment by jenniferarnesen** [comment]
Repo thoughts...

I am now leaning towards starting a new repo named `ui-widgets`, and gradually moving the components to that, rather than adding a ui-widgets package to d2-ui. In the end, it seems like we want to get away from the name d2-ui, so maybe the repo should disappear as well.

A different option would be for the `ui` repo to have 2 packages: `ui-core` and `ui-widgets`. Not sure...

Thoughts?

**Comment by Mohammer5** [comment]
@jenniferarnesen 
I think that's what @varl meant, I'd agree with this structure:

> rename:
> 
> * ui becomes ui-core
> * d2-ui becomes ui-widgets

I know that @varl wants to rename the `ui` repo to `ui-core`. But for the `d2-ui` I guess it makes sense to create a new repo?

**Comment by varl** [comment]
Yeah, or we let `dhis2/ui` stay that way and create `packages/ui-core` and `packages/ui-widgets` like @jenniferarnesen said.

I don't know why we keep getting pulled towards monorepos..

**Comment by ismay** [comment]
> I don't know why we keep getting pulled towards monorepos..

Monorepos, the silent killer.

**Comment by varl** [comment]
I guess people have a tendency to sort things into groups, and monorepos scratches that itch.

**Comment by varl** [comment]
> I am now leaning towards starting a new repo named `ui-widgets`, and gradually moving the components to that, rather than adding a ui-widgets package to d2-ui. In the end, it seems like we want to get away from the name d2-ui, so maybe the repo should disappear as well.

I think a new repo named `ui-widgets` would be OK. It makes automated releases easier at the very least. It also forces us to think about what goes into `ui-core` and it introduces some friction into adding new components which I think is good for the time being.

**Comment by jenniferarnesen** [comment]
Thinking about prerequisites for components that need d2. Before they can go into `ui-widgets`, I will suggest that we require that the component receives d2 as a property, and provides "drivers" (as described elsewhere by Viktor) for the supported api versions.

**Comment by ismay** [comment]
Shoud we also already think about how we want to eventually replace d2? Because eventually I would expect the components in widgets to just accept the props that they need (say, `user` or equivalent data), instead of being tied to the d2 api.

**Comment by varl** [comment]
> Thinking about prerequisites for components that need d2. Before they can go into `ui-widgets`, I will suggest that we require that the component receives d2 as a property, and provides "drivers" (as described elsewhere by Viktor) for the supported api versions.

I think we should postpone those components for now, or at least make a prototype and put it in a separate branch.

We do not want to head down into d2-ui territory again so we need to be cautious here.

**Comment by jenniferarnesen** [comment]
> I think we should postpone those components for now

There are already components needing d2 being prepared to go in, right @edoardo ?

**Comment by edoardo** [comment]
Yes, d2 is passed as prop.

**Comment by jenniferarnesen** [comment]
> Shoud we also already think about how we want to eventually replace d2? Because eventually I would expect the components in widgets to just accept the props that they need (say, `user` or equivalent data), instead of being tied to the d2 api.

Some of the components are pretty complex, like OrgUnitTree. I suspect that we are better off if that component handles the complex logic. Otherwise that logic would have to be repeated in the apps that use it.

**Comment by jenniferarnesen** [comment]
Here is a migration thought for the d2-ui components (that are not going to `ui-core`):

**d2-ui component**
  |   * get rid of side effects
  |   * use styled-jsx for styling
  |   * remove d2 as dependency and pass as property instead
  |   * "drivers" to handle agreed-on api versions
  |
  --> **d2-ui-widgets component**
               |
               |   * remove MUI dependency
               |
               --> **ui-widgets component**

Still, this does not address shared functions that are not components (like: https://github.com/dhis2/d2-ui-analytics/blob/master/src/api/dimensions.js)

**Comment by varl** [comment]
> ****
> 
> Still, this does not address shared functions that are not components (like: https://github.com/dhis2/d2-ui-analytics/blob/master/src/api/dimensions.js)

These look like they belong in [d2/analytics](https://github.com/dhis2/d2/tree/master/src/analytics)? I'm not sure what either is used for but they seem similar at a glance.
